### PR TITLE
METRON-1051: Enable the ability to update indexed messages

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
Currently our indexing layer is immutable. It would be useful to be able to:
* Update data in the indices
* Track lineage for updates

Current version of data should be able to be exposed via a batch interface. 
Updates should be available via a Java DAO API as well as via the REST API.
Implementation should ensure that the ability to read the current data in batch is possible.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by cestella** [comment]
Testing Instructions beyond the normal smoke test (i.e. letting data
flow through to the indices and checking them).

# Preliminaries

Set an environment variable to indicate `METRON_HOME`:
* `export METRON_HOME=/usr/metron/0.4.0` 

# Deploy the dummy parser
* Edit `$METRON_HOME/config/zookeeper/parsers/dummy.json`:
```
{
  "parserClassName":"org.apache.metron.parsers.json.JSONMapParser",
  "sensorTopic":"dummy"
}
```
* Create the dummy kafka topic:
  `/usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper node1:2181 --create --topic dummy --partitions 1 --replication-factor 1`
* Persist config changes: `$METRON_HOME/bin/zk_load_configs.sh -m PUSH -i $METRON_HOME/config/zookeeper -z node1:2181`
* Start via `$METRON_HOME/bin/start_parser_topology.sh -k node1:6667 -z node1:2181 -s dummy`

# Send dummy data through
* Edit `~/msg.json` with the following content:
```
{ "guid" : "guid0", "sensor.type" : "dummy", "timestamp" : 100 }
```
* Send `msg.json` through to kafka via `cat ~/msg.json | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list node1:6667 --topic dummy`

## Test Case: Update via patch
* Patch the message in ES and create a new field 'project' by executing
  the following:
```
curl -u user:password -X PATCH --header 'Content-Type: application/json' --header 'Accept: */*' -d '{
  "uuid" : "guid0",
    "sensorType" : "dummy",
    "patch" : [
    {
      "op": "add"
        , "path": "/project"
        , "value": "metron"
    }
  ]
}' 'http://node1:8082/api/v1/update/patch'
```
* Validate that the message has a field 'project':
```
curl -XPOST 'http://localhost:9200/dummy*/_search?pretty' -d '
{
  "_source" : [ "project" ]
}
'
```

## Test Case: Update via replace 
* Replace the message in ES and create a couple of modifications:
  * new field `new_field` == "brand new"
  * modified `timestamp` == 7
Execute the following:
```
curl -u user:password -X POST --header 'Content-Type: application/json' --header 'Accept: */*' -d '{
     "uuid" : "guid0",
     "sensorType" : "dummy",
     "replacement" : {
       "source:type": "dummy",
       "guid" : "guid0",
       "new_field" : "brand new",
       "timestamp" : 7
      }
   }' 'http://node1:8082/api/v1/update/replace'
```
* Validate that the message has a field 'project':
```
curl -XPOST 'http://localhost:9200/dummy*/_search?pretty' -d '
{
  "_source" : [ "new_field", "timestamp" ]
}
'
```
## Test Case: Validate that the changes persist to HBase
* Execute `echo "scan 'metron_update'" | hbase shell` 
* Ensure that 
  * There is one row: `guid0`
  * There are 2 columns 
  * The values associated with the columns correspond to the two modifications done in the tests above

**Comment by cestella** [comment]
Ok, architectural coverage is mentioned here, so I think this is ready for review.

**Comment by cestella** [comment]
Just a brief note about the `TableProvider` business.  We had cut and pasted a mock HTableProvider through several of our projects.  In order to test the `HBaseDao`, I needed yet another HTableProvider cut and pasted. I considered this a bridge too far. Instead of doing that, I
* Took the effort to create it once, put the implementation in the test artifact created by `metron-hbase`
* Refactor existing tests to use that consolidated implementation

I apologize for increasing the complexity of this PR, but I chose to interpret this as cleanup that testing this PR strongly suggested be done.

**Comment by justinleet** [comment]
Can you add Javadoc to the various interfaces and interface methods you added?  I know we've been inconsistent (i.e. we usually don't), but I think it'd be helpful as more people dig into things.

**Comment by cestella** [comment]
@justinleet yeah, I'll add javadoc to `IndexDao`

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java [hunk 1]
```diff
@@ -102,15 +128,100 @@ public SearchResponse search(SearchRequest searchRequest) throws InvalidSearchEx
...
+  }
+
+  @Override
+  public void update(Document update, Optional<String> index) throws IOException {
+    String indexPostfix = ElasticsearchUtils.getIndexFormat(accessConfig.getGlobalConfigSupplier().get()).format(new Date());
+    String sensorType = update.getSensorType();
+    String indexName = ElasticsearchUtils.getIndexName(sensorType, indexPostfix, null);
+
+    String type = sensorType + "_doc";
+    byte[] source = JSONUtils.INSTANCE.toJSON(update.getDocument());
```
Comment (justinleet) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: Why do we convert the update.getDocument() to a byte[], when the IndexRequest.source() method can take a Map?

Is there some correctness or typing concern?  I'm not particularly familiar with any subtleties of the ES apis like that.

Comment (cestella) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: I am honestly not sure about the ES API subtleties, but I'm mimicking [ElasticsearchWriter](https://github.com/apache/metron/blob/ec959d20efa72e9868d87a02d9407bbaad34c4c8/metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/ElasticsearchWriter.java#L88).

It seems to me that it's 6 of 1 and half a dozen of another considering it shouldn't matter terribly who converts to bytes, it's getting converted to bytes either way.

Comment (cestella) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: I just checked this in the tests, it seems to not make a difference one way or another.  I removed the conversion in favor of letting ES do it.

File: metron-platform/metron-indexing/src/main/java/org/apache/metron/indexing/dao/MultiIndexDao.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class MultiIndexDao implements IndexDao {
+  private List<IndexDao> indices;
+
+  public MultiIndexDao( IndexDao... composedDao) {
+    indices = new ArrayList<>();
+    for(IndexDao dao: composedDao) {
```
Comment (justinleet) [inline metron-platform/metron-indexing/src/main/java/org/apache/metron/indexing/dao/MultiIndexDao.java]: `Collections.addAll` again.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java [hunk 3]
```diff
@@ -102,15 +128,99 @@ public SearchResponse search(SearchRequest searchRequest) throws InvalidSearchEx
...
+
+  }
+
+  @Override
+  public void update(Document update, Optional<String> index) throws IOException {
+    String indexPostfix = ElasticsearchUtils.getIndexFormat(accessConfig.getGlobalConfigSupplier().get()).format(new Date());
+    String sensorType = update.getSensorType();
+    String indexName = ElasticsearchUtils.getIndexName(sensorType, indexPostfix, null);
+
+    String type = sensorType + "_doc";
```
Comment (cestella) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: I'm not sure, where would you expect that to be documented?  I'm relying on this component interacting with the writer components in a seamless manner, so there are some assumptions being made.  Perhaps I should make a comment here and denote that connection?

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchUpdateIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                , JSONUtils.INSTANCE.load(new String(columns.lastEntry().getValue())
+                        , new TypeReference<Map<String, Object>>() {})
+        );
+      }
+      {
+        //ensure ES is up-to-date
+        long cnt = 0;
+        for (int t = 0; t < MAX_RETRIES && cnt == 0; ++t, Thread.sleep(SLEEP_MS)) {
+          docs = es.getAllIndexedDocs(index, SENSOR_NAME + "_doc");
+          cnt = docs
```
Comment (cestella) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchUpdateIntegrationTest.java]: Good catch.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchUpdateIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        for (int t = 0; t < MAX_RETRIES && cnt == 0; ++t,Thread.sleep(SLEEP_MS)) {
+          docs = es.getAllIndexedDocs(index, SENSOR_NAME + "_doc");
+          cnt = docs
+                  .stream()
+                  .filter(d -> {
+                    Object newfield = d.get("new-field");
+                    return newfield != null && newfield.equals(message0.get("new-field"));
+                  }).count();
+        }
+        if (cnt == 0) {
```
Comment (cestella) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchUpdateIntegrationTest.java]: Yep, that's clearer.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java [hunk 6]
```diff
@@ -90,6 +95,38 @@ private void cleanDir(File dir) throws IOException {
         }
         dir.mkdirs();
     }
+
+    public BulkResponse add(String indexName, String sensorType, String... docs) throws IOException {
+        List<String> d = new ArrayList<>();
+        for(String doc : docs) {
```
Comment (cestella) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: Yep

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java (hunk 3 above)
Comment (cestella) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: Thinking about it for a minute, I think we need to create a `README.md` in `metron-elasticsearch` that details some of the assumptions for how messages are written in ES.  I'll create a JIRA around that.

Comment (cestella) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: I added [METRON-1080](https://issues.apache.org/jira/browse/METRON-1080).

Comment (justinleet) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: Thanks for creating that JIRA.  I think that's a good way to handle it.

Do we also need/want a corresponding doc for metron-solr?

Comment (cestella) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: Yeah, probably.

Comment (cestella) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java]: Ok, I made the JIRA [METRON-1082](https://issues.apache.org/jira/browse/METRON-1082)

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java [hunk 7]
```diff
@@ -43,10 +43,7 @@
 
 import java.io.File;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (JonZeolla) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: We should avoid this.

Comment (cestella) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: Right you are; corrected. :)

**Comment by cestella** [comment]
Alright, I think all of the concerns thus far are addressed.  Let me know if I missed anything @justinleet , et al

**Comment by justinleet** [comment]
Thanks for making the changes, I'm +1.

**Pull request closed without merging**